src/with_clause_parser.c

```diff
@@ -0,0 +1,164 @@
+ *         within_namespace: every element within the namespace
+ *     not_within_namespace: all the other elements
+ *
+ * That is, given a with clause like:
+ *     WITH (foo.foo_para, bar.bar_param, baz.baz_param)
+ *
+ * ts_with_clause_filter(elems, "foo", in, not_in) will have
+ *        in = foo.foo_para
+ *    not_in = bar.bar_param, baz.baz_param
+ */
+void
+ts_with_clause_filter(const List *def_elems, const char *namespace, List **within_namespace,
+					  List **not_within_namespace)
+{
+	ListCell *cell;
+
+	foreach (cell, def_elems)
+	{
+		DefElem *def = (DefElem *) lfirst(cell);
+
+		if (def->defnamespace != NULL && pg_strcasecmp(def->defnamespace, namespace) == 0)
+		{
+			if (within_namespace != NULL)
+				*within_namespace = lappend(*within_namespace, def);
+		}
+		else if (not_within_namespace != NULL)
+		{
+			*not_within_namespace = lappend(*not_within_namespace, def);
+		}
+	}
+}
+
+static Datum parse_arg(WithClauseArg arg, DefElem *def);
+
+/*
+ * Deserialize and apply the values in a WITH clause based on the on_arg table.
+ *
+ * This function will go through every element in def_elems and search for a
+ * corresponding argument in args, if one is found it will attemt to deserialize
+ * the argument, using that table elements deserialize function, then apply it
+ * to state.
+ *
+ * This is used to turn the list into a form more useful for our internal
+ * functions
+ */
+WithClauseParse
+ts_with_clauses_apply(const List *def_elems, const WithClauseArg *args, Size nargs)
+{
+	ListCell *cell;
+	bool *is_default = palloc(sizeof(*is_default) * nargs);
+	Datum *result = palloc0(sizeof(*result) * nargs);
+	Size i;
+	WithClauseParse parsed = {
+		.is_default = is_default,
+		.results = result,
+	};
+	memset(is_default, true, nargs);
+
+	for (i = 0; i < nargs; i++)
+		result[i] = args[i].default_val;
+
+	foreach (cell, def_elems)
+	{
+		DefElem *def = (DefElem *) lfirst(cell);
+		bool argument_recognized = false;
+
+		for (i = 0; i < nargs; i++)
+		{
+			if (pg_strcasecmp(def->defname, args[i].arg_name) == 0)
+			{
+				argument_recognized = true;
+
+				if (!is_default[i])
+					ereport(ERROR,
+							(errcode(ERRCODE_AMBIGUOUS_PARAMETER),
+								errmsg("duplicate parameter \"%s.%s\"",
+									def->defnamespace,
+									def->defname)));
+
+				result[i] = parse_arg(args[i], def);
+				is_default[i] = false;
+				break;
+			}
+		}
+
+		if (!argument_recognized)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unrecognized parameter \"%s.%s\"", def->defnamespace, def->defname)));
+	}
+
+	return parsed;
+}
+
+static Datum
+parse_arg(WithClauseArg arg, DefElem *def)
+{
+	char *value;
+	Datum val;
+	Oid in_fn;
+	Oid typIOParam;
+
+	if (!OidIsValid(arg.type_id))
+		elog(ERROR, "argument \"%s.%s\" not implemented", def->defnamespace, def->defname);
+
+	if (def->arg != NULL)
+		value = defGetString(def);
+	else
+		value = "true";
```

Comment:
`because `WITH (hypertable.some_bool_value)` is supposed to be equivalent to `WITH (hypertable.some_bool_value=true)`, and this is the way that postgres implemented it for their own WITH clause parser.  
Now that we have type info in the arg, think I should change it to only fire when `type_id == BOOLOID`? It would allow slightly better error messages for non-bool types, but would be less nice for bool-like types

Reply:
👍 to only doing this for bools

Reply:
also please add comment explaining this.

Reply:
ping